unblob/handlers/archive/zip.py

```diff
@@ -58,54 +65,49 @@ class ZIPHandler(StructHandler):
-        # the ZIP file chunk within our file independently, and then carve that chunk out.
-
-        file.seek(start_offset)
-
-        zip_end = find_first(file, EOCD_RECORD_HEADER)
-
-        if zip_end == -1:
-            raise InvalidInputFormat("Missing EOCD record header in ZIP chunk.")
-
-        file.seek(zip_end)
-        self.cparser_le.end_of_central_directory_t(file)
-        return file.tell()
-
-    def check_file(self, file: io.BufferedIOBase) -> Tuple[bool, bool]:
+    def has_encrypted_files(
+        self,
+        file: io.BufferedIOBase,
+        start_offset: int,
+        end_of_central_directory: Instance,
+    ) -> bool:
         has_encrypted_files = False
-        try:
-            with zipfile.ZipFile(file) as zip:  # type: ignore
-                for zipinfo in zip.infolist():
-                    if zipinfo.flag_bits & ENCRYPTED_FLAG:
-                        has_encrypted_files = True
-            return True, has_encrypted_files
-        except (zipfile.BadZipFile, UnicodeDecodeError, ValueError):
-            return False, has_encrypted_files
+        file.seek(start_offset + end_of_central_directory.offset_of_cd)
+        for _ in range(0, end_of_central_directory.total_entries):
+            cd_header = self.cparser_le.cd_file_header_t(file)
+            if cd_header.flags & ENCRYPTED_FLAG:
+                has_encrypted_files = True
+        return has_encrypted_files
 
     def calculate_chunk(
         self, file: io.BufferedIOBase, start_offset: int
     ) -> Optional[ValidChunk]:
 
-        header = self.parse_header(file)
-        if header.version_needed_to_extract > MAXIMUM_VERSION:
-            return
-
-        end_of_zip = self._calculate_zipfile_end(file, start_offset)
-        file.seek(start_offset)
+        has_encrypted_files = False
+        zip_end = -1
+        file.seek(start_offset, io.SEEK_SET)
+
+        for offset in iterate_patterns(file, struct.pack("<I", EOCD_RECORD_HEADER)):
+            file.seek(offset)
+            end_of_central_directory = self.cparser_le.end_of_central_directory_t(file)
+            central_directory_offset = (
+                start_offset + end_of_central_directory.offset_of_cd
+            )
+            file.seek(central_directory_offset)
+            magic = convert_int32(file.read(4), Endian.LITTLE)
+            if magic == CD_RECORD_HEADER:
+                zip_end = offset
+                has_encrypted_files = self.has_encrypted_files(
+                    file, start_offset, end_of_central_directory
+                )
+                break
 
-        is_valid, has_encrypted_files = self.check_file(file)
+        if zip_end == -1:
+            raise InvalidInputFormat("Missing EOCD record header in ZIP chunk.")
 
-        if not is_valid:
-            raise InvalidInputFormat("Invalid ZIP header.")
+        file.seek(zip_end)
+        self.cparser_le.end_of_central_directory_t(file)
+        end_of_zip = file.tell()
```

Comment:
maybe do the above check here?

Reply:
Not entirely sure of what you mean here.